src/main/java/com/google/api/gax/grpc/RetryingCallable.java

```diff
@@ -116,58 +125,147 @@ private Retryer(
-                Retryer retryer =
-                    new Retryer(request, context, result, retryDelay, rpcTimeout, throwable);
-                executor.schedule(
-                    retryer, DEADLINE_SLEEP_DURATION.getMillis(), TimeUnit.MILLISECONDS);
-                return;
-              }
-
-              long newRetryDelay =
-                  (long) (retryDelay.getMillis() * retryParams.getRetryDelayMultiplier());
-              newRetryDelay = Math.min(newRetryDelay, retryParams.getMaxRetryDelay().getMillis());
-
-              long newRpcTimeout =
-                  (long) (rpcTimeout.getMillis() * retryParams.getRpcTimeoutMultiplier());
-              newRpcTimeout = Math.min(newRpcTimeout, retryParams.getMaxRpcTimeout().getMillis());
-
-              long randomRetryDelay = ThreadLocalRandom.current().nextLong(retryDelay.getMillis());
-              Retryer retryer =
-                  new Retryer(
-                      request,
-                      context,
-                      result,
-                      Duration.millis(newRetryDelay),
-                      Duration.millis(newRpcTimeout),
-                      throwable);
-              executor.schedule(retryer, randomRetryDelay, TimeUnit.MILLISECONDS);
-            }
-          });
+      resultFuture.issueCall(request, deadlineContext, this);
+    }
+
+    @Override
+    public void onSuccess(ResponseT r) {
+      resultFuture.set(r);
+    }
+
+    @Override
+    public void onFailure(Throwable throwable) {
+      if (!canRetry(throwable)) {
+        resultFuture.setException(throwable);
+        return;
+      }
+      if (isDeadlineExceeded(throwable)) {
+        Retryer retryer =
+            new Retryer(request, context, resultFuture, retryDelay, rpcTimeout, throwable);
+        resultFuture.scheduleNext(
+            executor, retryer, DEADLINE_SLEEP_DURATION.getMillis(), TimeUnit.MILLISECONDS);
+        return;
+      }
+
+      long newRetryDelay = (long) (retryDelay.getMillis() * retryParams.getRetryDelayMultiplier());
+      newRetryDelay = Math.min(newRetryDelay, retryParams.getMaxRetryDelay().getMillis());
+
+      long newRpcTimeout = (long) (rpcTimeout.getMillis() * retryParams.getRpcTimeoutMultiplier());
+      newRpcTimeout = Math.min(newRpcTimeout, retryParams.getMaxRpcTimeout().getMillis());
+
+      long randomRetryDelay = ThreadLocalRandom.current().nextLong(retryDelay.getMillis());
+      Retryer retryer =
+          new Retryer(
+              request,
+              context,
+              resultFuture,
+              Duration.millis(newRetryDelay),
+              Duration.millis(newRpcTimeout),
+              throwable);
+      resultFuture.scheduleNext(executor, retryer, randomRetryDelay, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private class RetryingResultFuture extends AbstractFuture<ResponseT> {
+    private volatile Future<?> activeFuture = null;
+    private volatile boolean cancelled = false;
```

Comment:
I don't think we need this. AbstractFuture already has the cancel()-isCancelled() pair right?